src/plugins.c

```diff
@@ -562,6 +705,8 @@ plugin_free(Plugin *plugin)
 
 	plugin_list = g_list_remove(plugin_list, plugin);
 
+	if (plugin->cb_data_destroy)
+		plugin->cb_data_destroy(plugin->cb_data);
```

Comment:
You seem to be using it wrong (your description sounds right though). You can use geany_plugin_set_data() in geany_load_module(), even if the plugin is never enabled by the user. What did the free_func access? It should really just free what you've previously allocated in geany_load_module().

It doesn't have the same semantics as cleanup, don't use it as a drop-in replacement.